### DISCUSSION TITLE
ASR: disable progress-bar for transcription?

### DISCUSSION THREAD

**Discussion by domcross** [NVIDIA-NeMo/NeMo#3281] [Q&A] [answered]
Hello all,

probably I am missing something obvious - is it possible to disable the progress-bar (tqdm) for the `transcribe`-method of an asr-model?

Thanks for any suggestions in advance.

Dominik

**Comment by titu1994** [comment]
Not particularly, is there a usecase for removing the progress bar?

**Comment by domcross** [comment]
I want to run a custom evaluation script with my own pre/post-processing where I have set batchsize to "1".
Currently I have two progress bars (one from asr-model.transcribe and one from my sourrounding processing loop) which overwrite each other...

**Reply by titu1994** [reply]
Removing the progress bar will be tricky. Nested tqdm should work fine though. If we do need to support this it will be a change in too many files and make the transcribe() method have to accept another argument verbose=True which would not be great since in nearly all cases you want that to be true

**Reply by prakhark28** [reply]
Adding a verbose argument makes sense as almost every time one uses the transcribe function, it is inside an outer loop with tqdm. The transcribe bar messes up the output logs.

**Comment by edStorey** [comment, marked as answer]
I have also had the same problem as @domcross, I am looping through several files using the `transcribe` function with a batch size of 1 similar to the above.

I used this fix [codereview.stackexchange.com/a/25471](https://codereview.stackexchange.com/a/25471) to get around the issue by running the `transcribe` function in a with statement with their `NoStdStreams` class this stopped the progress bar from appearing. This was not ideal however, functionality to disable progress bars would be appreciated in future nemo releases.

**Reply by prakhark28** [reply]
```
for audio in tqdm(data):    
    with NoStdStreams():
        logits.append(asr_model.transcribe([audio['audio_filepath']], logprobs=True)[0])
```
This returns a tqdm and transcribe progress bar alternately. Can you help me mute the transcribe outputs

**Reply by edStorey** [reply]
```
class NoStdStreams(object):
        def __init__(self,stdout = None, stderr = None):
            self.devnull = open(os.devnull,'w')
            self._stdout = stdout or self.devnull or sys.stdout
            self._stderr = stderr or self.devnull or sys.stderr

        def __enter__(self):
            self.old_stdout, self.old_stderr = sys.stdout, sys.stderr
            self.old_stdout.flush(); self.old_stderr.flush()
            sys.stdout, sys.stderr = self._stdout, self._stderr

        def __exit__(self, exc_type, exc_value, traceback):
            self._stdout.flush(); self._stderr.flush()
            sys.stdout = self.old_stdout
            sys.stderr = self.old_stderr
            self.devnull.close() 
```

That is the NoStdStreams class I used  ^^

```
with NoStdStreams() : 
    transcription = asr_model.cuda().transcribe(paths2audio_files=file_path)
```

And that is my implementation, performing the transcription on gpu with `cuda()`.

Hope that helps!

**Comment by evilc3** [comment]
I used these lines to disable tqdm completely 

`from tqdm import tqdm 

from functools import partialmethod 

tqdm.__init__ = partialmethod(tqdm.__init__, disable=True)
`

**Comment by tylercasper** [comment]
This is an old topic, but when streaming, it is very desirable to have the progress and inference rate output be suppressed. With an 80 ms lookahead and 80 ms encoder step length, this output is printed excessively.
